Updates to tileslist, list and marquee selection

wiolsen opened the pull request:
#### Pull request checklist
#### Description of changes
- List.tsx - Limit number of calls to getBoundingClientRect in to improve perf
- List.tsx - Pass ItemsPerPage to get page height function in 
- List.tsx - add getTotalListHeight() method
- List.Types.ts - add 'isVisible' parameter to IPage so that list owner can know which pages are visible to the user
- MarqueeSelection.tsx - limit selection change updates to only call when the selection actually updates for perf improvement
- MarqeeSelection.tsx - constrain the dragRect to be within the viewport
- TilesList.tsx - Implementing proper support for cellPerPage prop override
- TilesList.tsx - added scrollToIndex method
- TilesList.tsx - added ListProps props so that owner can pass props through to the underlying list






#### Focus areas to test


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9439)

KevinTCoughlin commented on packages/office-ui-fabric-react/src/components/List/List.tsx:
```diff
@@ -304,6 +309,10 @@ export class List<T = any> extends BaseComponent<IListProps<T>, IListState<T>> i
   }
 
+  public componentWillUnmount(): void {
+    this._events.dispose();
```
`BaseComponent` handles this for you already

https://github.com/OfficeDev/office-ui-fabric-react/blob/de80247f560fb65a97d06873a0678f76f7f9744d/packages/utilities/src/BaseComponent.ts#L85

wiolsen replied:
I removed this

KevinTCoughlin commented on packages/office-ui-fabric-react/src/components/List/List.tsx:
```diff
@@ -552,6 +572,7 @@ export class List<T = any> extends BaseComponent<IListProps<T>, IListState<T>> i
    * This function is debounced.
    */
   private _onScrollingDone(): void {
+    this._onAsyncIdle(); // Guarantees the creation of lookahead images when scrolling is done
```
This may cause perf regressions, we'll need to look at the profiler. `_onAsyncIdle` is expensive IIRC.

wiolsen replied:
I did some perf tests and it seems OK to remove this.  I am brining over some changes made by other developers, so I am not sure why this was necessary.  I have removed it.  Thank you.

KevinTCoughlin commented on packages/experiments/src/components/TilesList/TilesList.tsx:
```diff
@@ -105,13 +108,13 @@ export class TilesList<TItem> extends React.Component<ITilesListProps<TItem>, IT
         role={role}
         {...divProps}
-        ref={ref as ((element: FocusZone | null) => void)}
+        ref={focusZoneComponentRef as ((element: FocusZone | null) => void)}
```
This will break our API contract unless I'm misunderstanding. The above is replacing a `HTMLElement` ref with a `componentRef` prop set on the following line:

https://github.com/OfficeDev/office-ui-fabric-react/pull/9439/files#diff-533bdd4a04491852153288ac46951998R118

KevinTCoughlin commented on packages/office-ui-fabric-react/src/components/List/List.types.ts:
```diff
@@ -155,6 +160,13 @@ export interface IListProps<T = any> extends React.HTMLAttributes<List<T> | HTML
 
+  /**
+   * Boolean to disable the onAsyncIdle call. Recommended to use when doing jumping through the list.
+   * As the user is moving through the list very quickly, creating lookaheads is heavy and unneccessary.
```
_It depends_. The lookahead calculations should stop once the entire set of pages have been traversed. The only time that this _does not_ occur is when the `window` is the scrollable viewport. 

https://github.com/OfficeDev/office-ui-fabric-react/issues/8740

Can you explain your scenario a bit more? We can discuss offline.

wiolsen replied:
I removed this change.

KevinTCoughlin commented on packages/office-ui-fabric-react/src/components/MarqueeSelection/MarqueeSelection.base.tsx:
```diff
@@ -74,6 +75,10 @@ export class MarqueeSelectionBase extends BaseComponent<IMarqueeSelectionProps,
+    if (this._evaluateSelection) {
+      delete this._evaluateSelection;
+    }
+    this._events.dispose();
```
Not necessary since this component extends `BaseComponent`. See my similar comment https://github.com/OfficeDev/office-ui-fabric-react/pull/9439/files#r293216112.

wiolsen replied:
I removed this.

KevinTCoughlin commented on packages/experiments/src/components/TilesList/TilesList.tsx:
```diff
@@ -122,11 +125,37 @@ export class TilesList<TItem> extends React.Component<ITilesListProps<TItem>, IT
   }
 
+  public scrollToIndex(index: number, mode: ScrollToMode = ScrollToMode.auto): void {
+    if (this.listRef && this.listRef.current) {
```
Nit: you can probably refactor out the `this.listRef` portion of the conditional when assigning the member field

wiolsen replied:
removed the this.listRef in the conditional

KevinTCoughlin commented on packages/experiments/src/components/TilesList/TilesList.tsx:
```diff
@@ -351,7 +380,10 @@ export class TilesList<TItem> extends React.Component<ITilesListProps<TItem>, IT
-    const endIndex = Math.min(cells.length, startIndex + CELLS_PER_PAGE);
+    const endIndex = Math.min(
+      cells.length,
+      this.props.cellsPerPage && this.props.cellsPerPage > 0 ? startIndex! + this.props.cellsPerPage : startIndex! + CELLS_PER_PAGE
```
Is the non-null assertion necessary? It wasn't there prior.

wiolsen replied:
I removed the non null assertions.  We were running our code with different linter setting that required that to be there.

KevinTCoughlin commented on packages/office-ui-fabric-react/src/components/List/List.tsx:
```diff
@@ -164,6 +168,7 @@ export class List<T = any> extends BaseComponent<IListProps<T>, IListState<T>> i
     this._estimatedPageHeight = 0;
     this._focusedIndex = -1;
     this._pageCache = {};
+    this.refs = {};
```
Unused?

wiolsen replied:
Removed. This was also put there to fix linter issues due to stricter linter settings in our build.  But, it is not needed here.

ThomasMichon commented on packages/experiments/src/components/TilesList/TilesList.types.ts:
```diff
@@ -123,4 +124,13 @@ export interface ITilesListProps<TItem> extends IBaseProps, React.Props<TilesLis
+  /**
+   * calback to return the height of a single item in pixels
+   */
+  getItemHeight?: (index: number) => number;
```
It seems like `getItemHeight` would conflict with `minRowHeight` and `desiredSize`. What item does this apply to?

wiolsen replied:
I removed this callback prop in the latest iteration.

wiolsen replied:
"Item" is a "cell" in this case.  This should be made clear in the latest revision.

ThomasMichon commented on packages/experiments/src/components/TilesList/TilesList.tsx:
```diff
@@ -122,11 +125,37 @@ export class TilesList<TItem> extends React.Component<ITilesListProps<TItem>, IT
+        index,
+        (itemIndex: number) => {
+          let cellHeight = 0;
+          if (this.props.getItemHeight) {
```
Shouldn't `TilesList` itself be able to calculate the offset from its grid data?

wiolsen replied:
yes it can :)  so, I have removed the getItemHeight callback in the latest PR iteration.  Thank you

ThomasMichon commented on packages/experiments/src/components/TilesList/TilesList.tsx:
```diff
@@ -351,7 +380,10 @@ export class TilesList<TItem> extends React.Component<ITilesListProps<TItem>, IT
     const { cells } = this.state;
 
-    const endIndex = Math.min(cells.length, startIndex + CELLS_PER_PAGE);
+    const endIndex = Math.min(
```
```ts
const {
  cellsPerPage = CELLS_PER_PAGE
} = this.props;
```
...
```
const endIndex = Math.min(cells.length, startIndex + cellsPerPage);
```

wiolsen replied:
fixed in latest iteration.

aneeshack4 commented:
pinging @wiolsen looks like there's some more feedback to address!

KevinTCoughlin commented:
@wiolsen there's a initial render regression when I was comparing existing behavior with this PR. Compare:

- Before changes: http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/9404/merge/fabric-website-resources/dist/index.html#/tests/detailslistbasicexample
- After changes: http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/9439/merge/fabric-website-resources/dist/index.html#/tests/detailslistbasicexample

**Before**

![before_list_changes](https://user-images.githubusercontent.com/706967/59791792-b8317500-9287-11e9-9cc4-edceb72215fa.gif)

**After**

![after_list_changes](https://user-images.githubusercontent.com/706967/59791804-bbc4fc00-9287-11e9-92d8-ebba7d6fe2fe.gif)


It looks measurement related. With the changes the initial render seemed slower, but I didn't look at a profiler snapshot. Can you take a look?

wiolsen commented:
@KevinTCoughlin I think I have fixed the apparent performance issue, and in the process I found and fixed another issue with the marquee selection in /#/tests/detailslistbasicexample.

I have made the changes, and when I run /#/tests/detailslistbasicexample locally, everything seems good there.

KevinTCoughlin commented:
Spoke offline, there's an issue with paging only repro'ing on http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/9439/merge/fabric-website-resources/dist/index.html#/examples/detailslist

![image](https://user-images.githubusercontent.com/706967/59806987-30f5f880-92ab-11e9-8be0-71e83b46662a.png)

@wiolsen is looking into it.

wiolsen commented:
I decided to remove the changes that were causing issues in details list example.  Hopefully, I can revisit those fixes to scrolling performance in a future PR.

KevinTCoughlin commented on common/changes/office-ui-fabric-react/wiolsen-pr-changes_2019-06-18-22-39.json:
```diff
@@ -0,0 +1,11 @@
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Updates to tileslist, list and marquee selection",
```
```suggestion
      "comment": "List: Add public method to return total list height",
```

KevinTCoughlin commented on common/changes/@uifabric/experiments/wiolsen-pr-changes_2019-06-18-22-39.json:
```diff
@@ -0,0 +1,11 @@
+  "changes": [
+    {
+      "packageName": "@uifabric/experiments",
+      "comment": "Updates to tileslist, list and marquee selection",
```
```suggestion
      "comment": "Add public methods for scroll to index and return total list height to TilesList. Improve MarqueeSelection behavior.",
```

dzearing commented on packages/office-ui-fabric-react/src/components/MarqueeSelection/MarqueeSelection.base.tsx:
```diff
@@ -216,22 +217,52 @@ export class MarqueeSelectionBase extends BaseComponent<IMarqueeSelectionProps,
         }
+
         // We need to constrain the current point to the rootRect boundaries.
-        const constrainedPoint = this.props.isDraggingConstrainedToRoot
```
It seems like `isDraggingConstrainedToRoot` might not be respected any more. Did this get dropped?

wiolsen replied:
It turns out that isDraggingConstrainedToRoot was working previously.  I think that the person who made the change on our end was not aware of this prop, so they rewrote / broke the behavior.  Sorry I did not catch this, and thank your for noticing.

wiolsen replied:
So, I have removed our changes related to constraining the drag rect.  It meets our needs to just set isDragginConstrainedToRoot to true.

dzearing commented on packages/office-ui-fabric-react/src/components/MarqueeSelection/MarqueeSelection.base.tsx:
```diff
@@ -216,22 +217,52 @@ export class MarqueeSelectionBase extends BaseComponent<IMarqueeSelectionProps,
+          y: this._lastMouseEvent!.clientY - rootRect.top
+        };
+
+        let rectLeft = Math.min(this._dragOrigin.x, constrainedPoint.x);
```
super ultra minor nit, if you just call this `left`, and `rectTop` just `top`, you can later just use them like:

`const dragRect = { left, top, width, height };`

wiolsen replied:
fixed

dzearing commented on packages/office-ui-fabric-react/src/components/MarqueeSelection/MarqueeSelection.base.tsx:
```diff
@@ -216,22 +217,52 @@ export class MarqueeSelectionBase extends BaseComponent<IMarqueeSelectionProps,
+          }
+        }
 
         const dragRect = {
```
i'm not totally sure I follow the before/after behavioral change here. Maybe you could add a test to illustrate what behavior didn't work before that now works.

wiolsen replied:
I have removed this change.

dzearing commented on packages/office-ui-fabric-react/src/components/MarqueeSelection/MarqueeSelection.base.tsx:
```diff
@@ -360,18 +387,53 @@ export class MarqueeSelectionBase extends BaseComponent<IMarqueeSelectionProps,
+    const previousSelectedIndices = this._allSelectedIndices;
+    this._allSelectedIndices = {};
+
+    // set all indices that is supposed to be selected in _allSelectedIndices
```
// that *are* supposed to be selected

wiolsen replied:
fixed

dzearing commented on packages/office-ui-fabric-react/src/components/MarqueeSelection/MarqueeSelection.base.tsx:
```diff
@@ -360,18 +387,53 @@ export class MarqueeSelectionBase extends BaseComponent<IMarqueeSelectionProps,
+
+    // check if needs to update selection, only when current _allSelectedIndices
+    // is different than previousSelectedIndices
+    let needToUpdate = true;
```
Wait. Shouldn't this default to false?

dzearing replied:
Please add a test to catch this bug. `needToUpdate` is never false.

wiolsen replied:
oh.. glad you caught that.  I was doing some tests on a previous PR iteration where I was forcing it to be true while diagnosing another problem, but forgot to set it back.  I am working on adding a test now.

wiolsen replied:
I have added a simple unit test that verifies that making a marquee around a selectable element will update the selection.  This will catch whether needToUpdate was properly set.

dzearing commented on packages/office-ui-fabric-react/src/components/List/List.tsx:
```diff
@@ -358,6 +361,17 @@ export class List<T = any> extends BaseComponent<IListProps<T>, IListState<T>> i
+   * Get the current height the list and it's pages.
+   */
+  public getTotalListHeight(): number {
+    let height = 0;
```
Rather than iterating, wouldn't `_surfaceRect.height` fit the purpose?

wiolsen replied:
Yes, that seems to work.  I just did some tests to confirm that they return the same value.  I will put the change in the next iteration.

wiolsen replied:
I have put this change in.

dzearing commented on packages/office-ui-fabric-react/src/components/List/List.tsx:
```diff
@@ -779,6 +793,9 @@ export class List<T = any> extends BaseComponent<IListProps<T>, IListState<T>> i
         newPage.top = pageTop;
         newPage.height = pageHeight;
+        if (this._visibleRect && this._visibleRect.bottom) {
+          newPage.isVisible = pageBottom >= this._visibleRect.top && pageTop <= this._visibleRect.bottom;
```
is this used anywhere?

wiolsen replied:
Yes, we need this for a feature in odsp.
